### ISSUE TITLE
6.6: changed behaviour on pressing Enter on a selected block

### ISSUE THREAD

**Issue by elzadj** [WordPress/gutenberg#65274]
### Description

Hi everyone! 

We noticed that hitting Enter on our custom blocks inserts an empty paragraph since the ~~6.6.2~~ 6.6 update. Before ~~6.6.2~~ 6.6 it was inserting a copy of the same block.

~~We suspect that this has something to do with that fix: https://github.com/WordPress/gutenberg/pull/64992/files~~

The block in question: https://github.com/maxi-blocks/maxi-blocks/tree/master/src/blocks/text-maxi

Can you advise something, please? Maybe we can make some changes to revert it without rolling back to 6.5.5? Something for splitting the block, possibly?

Note: Adding support for "splitting": doesn't help.

Thanks!

### Step-by-step reproduction instructions

1. https://wordpress.org/plugins/maxi-blocks/ 
2. Add a Container + Text inside a Column
3. Press Enter on the Text, Column, Row.
Or check the videos below.

### Screenshots, screen recording, code snippet

Before 6.6:

[screencast-maxiblocks.com-2024.09.12-12_54_32.webm](https://github.com/user-attachments/assets/ebc77962-e41c-4b49-87e3-5451ca8483f8)

On 6.6.2:

[screencast-wordpress.local-2024.09.12-13_23_08 (1).webm](https://github.com/user-attachments/assets/854900ab-8a11-479d-86a1-10f27f45af32)

### Environment info

WordPress 6.6.2
PHP 8.3

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [X] Yes

**Comment by Mamaduka** [comment]
cc @stokesman

**Comment by stokesman** [comment]
So far I’m not able to reproduce the desired behavior on 6.6.1. I find that, from a Text Maxi block, pressing <kbd>Enter</kbd> inserts a core paragraph, whether the the block is nested in a Container Maxi block or not. That was testing MaxiBlocks 1.9.5.

I notice in the "Before 6.6.2" screen recording there seem to be more plugins enabled than in the other screen recording. Also I don’t see the Document Bar. It has been around since before 6.6.1 so it makes me wonder what the actual version of WP is in that screen recording.

<img width="978" alt="document-bar" src="https://github.com/user-attachments/assets/390ba34d-7a73-41c5-adfc-41cd791ed60e">

☝️ Document Bar for clarity’s sake.


> We suspect that this has something to do with that fix: https://github.com/WordPress/gutenberg/pull/64992/files

It doesn’t seem likely to be related. Those changes and that whole file are solely involved with handling a mouse event and I don’t see how that would affect this keyboard behavior.

**Comment by Mamaduka** [comment]
This could be related to the [`onSplit` callback deprecation](https://github.com/WordPress/gutenberg/pull/54543) in WP 6.6. I see that the Text Moxi block is [using it](https://github.com/maxi-blocks/maxi-blocks/blob/ebe9193d340dad6e3fca22e35079fe4d4246e52b/src/blocks/text-maxi/edit.js#L182-L189).

This was mentioned in [dev note](https://make.wordpress.org/core/2024/06/24/miscellaneous-editor-changes-in-wordpress-6-6/#Unified-split-logic-for-writing-flow) as well.

**Comment by elzadj** [comment]
> This could be related to the [`onSplit` callback deprecation](https://github.com/WordPress/gutenberg/pull/54543) in WP 6.6. I see that the Text Moxi block is [using it](https://github.com/maxi-blocks/maxi-blocks/blob/ebe9193d340dad6e3fca22e35079fe4d4246e52b/src/blocks/text-maxi/edit.js#L182-L189).
> 
> This was mentioned in [dev note](https://make.wordpress.org/core/2024/06/24/miscellaneous-editor-changes-in-wordpress-6-6/#Unified-split-logic-for-writing-flow) as well.

Yes, it definitely started from 6.6, and is connected to onSplit deprecation. 

But it seems like we can't re-create the behaviour we had before 6.6? We want the block to support splitting, but we also want to create a new empty Text block when we hit Enter at the end of it. It looks like [creating an empty paragraph is the default now ](https://github.com/WordPress/gutenberg/pull/54543#issuecomment-2200974976), right?

It's [also mentioned that onSplit should continue to work](https://github.com/WordPress/gutenberg/pull/54543#issuecomment-2202087837), but seems like it's not the case any more. 

Is there any alternative to onSplit?

**Comment by joshualip-plaudit** [comment]
I tried to implement a block that I need to split off into a new heading today, and encountered the raft of issues you've found here.
Based on the digging through the code I've done, the answer is as maddening as it is simple: All functionality related to onSplit has been replaced with... something that has yet to be properly documented.
As far as I can tell, the proposed, "unified flow" devolved into, "just shove a paragraph in if the split happened at the end and duplicate the block otherwise" (which, obviously, is a poor replacement for onSplit).

**However**, none of that is what actually got implemented. As far as I can tell, the final functionality is currently remarkably buggy (#64071, #64771) and the actual outcome appears to be improperly dependent on metadata properties. In addition, support for onSplit was completely removed in direct contradiction to [the PR comment you referenced](https://github.com/WordPress/gutenberg/pull/54543#issuecomment-2202087837). This is, of course, also in direct contradiction to the documentation on WordPress' own website.

**Comment by ellatrix** [comment]
> Is there any alternative to onSplit?

Yes, set `splitting` to `true` in your block supports object. See https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#splitting

> It looks like https://github.com/WordPress/gutenberg/pull/54543#issuecomment-2200974976, right?

No, the default is whatever is configured as the default block. See https://developer.wordpress.org/block-editor/reference-guides/packages/packages-blocks/#setdefaultblockname



Let me know if something doesn't work, maybe with minimal block registration code to reproduce.

**Comment by stokesman** [comment]
> It's [also mentioned that onSplit should continue to work](https://github.com/WordPress/gutenberg/pull/54543#issuecomment-2202087837), but seems like it's not the case any more.

> In addition, support for onSplit was completely removed in direct contradiction to [the PR comment you referenced](https://github.com/WordPress/gutenberg/pull/54543#issuecomment-2202087837).

I was just testing this out with a block I’m developing and concur. `onSplit` has ceased to work. I found that adding the [`identifier`](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/rich-text#identifier-string) prop to `RichText` makes it work again (though the `onSplit` callback isn’t invoked so I guess it’s just a non-falsy value at this point). To restate, `identifier` being required for `onSplit` to work is new—so the change was breaking. In my case it works to use the new `splitting` support and not a big deal.

**Comment by stokesman** [comment]
Also the reason `splitting` support works fine in my case is because the block to split is the only allowed block type in the parent. That means, even if the split is at the end, the same block type is used—not the default block. I can see how it is problematic that a split at the end would otherwise insert the default block. It doesn’t seem right that to have that work you’d need to use `setDefaultBlockName` since that’s the editor level setting.

**Comment by gchtr** [comment]
I can confirm that `onSplit` stopped working for us completely. When switching to `splitting` support, it only worked once we added the `identifier` property to the `RichText` component.

However, this only works if there’s an attribute name we can pass to the `identifier` property. It doesn’t work if the block takes the value from a source other than attributes. In our case, it’s a meta field, where we load the value through `useEntityProp()`. In this case, hitting enter does nothing.

**Comment by albanyacademy** [comment]
> > It's [also mentioned that onSplit should continue to work](https://github.com/WordPress/gutenberg/pull/54543#issuecomment-2202087837), but seems like it's not the case any more.
> 
> > In addition, support for onSplit was completely removed in direct contradiction to [the PR comment you referenced](https://github.com/WordPress/gutenberg/pull/54543#issuecomment-2202087837).
> 
> I was just testing this out with a block I’m developing and concur. `onSplit` has ceased to work. I found that adding the [`identifier`](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/rich-text#identifier-string) prop to `RichText` makes it work again (though the `onSplit` callback isn’t invoked so I guess it’s just a non-falsy value at this point). To restate, `identifier` being required for `onSplit` to work is new—so the change was breaking. In my case it works to use the new `splitting` support and not a big deal.

i didnt know about that identifier prop requirement change, good to know. i'll have to see if that fixes my issue #64071
